src/parser/jobs/mnk/Forms.js

```diff
@@ -0,0 +1,191 @@
+import React, {Fragment} from 'react'
+
+import {ActionLink, StatusLink} from 'components/ui/DbLink'
+import ACTIONS, {getAction} from 'data/ACTIONS'
+import STATUSES from 'data/STATUSES'
+import Module from 'parser/core/Module'
+import RAISE_STATUSES from 'parser/core/modules/Death'
+import {Suggestion, SEVERITY} from 'parser/core/modules/Suggestions'
+
+const FORM_TIMEOUT_MILLIS = 10000
+
+const FORMS = [
+	STATUSES.OPO_OPO_FORM.id,
+	STATUSES.RAPTOR_FORM.id,
+	STATUSES.COEURL_FORM.id,
+]
+
+const OPO_OPO_SKILLS = [
+	ACTIONS.BOOTSHINE.id,
+	ACTIONS.DRAGON_KICK.id,
+	ACTIONS.ARM_OF_THE_DESTROYER.id,
+]
+
+export default class Forms extends Module {
+	static handle = 'forms'
+	static dependencies = [
+		'death',
+		'downtime',
+		'combatants',
+		'suggestions',
+	]
+
+	_formless = 0
+	_poorForms = 0
+	_resetForms = 0
+	_skippedForms = 0
+	_droppedForms = 0
+
+	_lastFormChanged = null
+	_lastFormDropped = null
+
+	constructor(...args) {
+		super(...args)
+		this.addHook('cast', {by: 'player'}, this._onCast)
+		this.addHook('applybuff', {to: 'player', abilityId: FORMS}, this._onGain)
+		this.addHook('removebuff', {to: 'player', abilityiD: FORMS}, this._onRemove)
+		this.addHook('complete', this._onComplete)
+	}
+
+	_onCast(event) {
+		const action = getAction(event.ability.guid)
+
+		if (!action) {
+			return
+		}
+
+		if (action.onGcd) {
+			// Check the current form and stacks, or zero for no form
+			const current_form = FORMS.find(form => this.combatants.selected.hasStatus(form)) || 0
+			const untargetable = this.downtime.getDowntime(this._lastFormChanged || 0, event.timestamp)
+
+			// If PB is active, we can ignore form unless someone is really derpy
+			if (this.combatants.selected.hasStatus(STATUSES.PERFECT_BALANCE.id)) {
+				if (action === ACTIONS.FORM_SHIFT.id) {
+					this._poorForm++
+				}
+				return
+			}
+
+			// If we're in downtime or had poor HP uptime, ignore Form Shift
+			if (action === ACTIONS.FORM_SHIFT.id) {
+				// Recovery
+				if (RAISE_STATUSES.find(dead => this.combatants.select.hasStatus(dead))) {
+					return
+				}
+
+				// Downtime
+				if (untargetable !== 0) { return }
+
+				// Actually using Form Shift
+				this._skippedForms++
+
+				return
+			}
+
+			// Handle relevant actions per form
+			switch (current_form) {
+			case STATUSES.OPO_OPO_FORM.id:
+				break
+			case STATUSES.RAPTOR_FORM.id:
+				// Using Opo-Opo skills resets form
+				if (OPO_OPO_SKILLS.includes(action)) { this._resetForms++ }
+				break
+			case STATUSES.COEURL_FORM.id:
```

Comment:
Can probably afford to let the switch fall through for these two, saves duping the handling.